orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+        - Extracts relevant trainer options
+        - Extracts partition information from the training session in case of a ZeRO enabled distributed
+            ORTTrainer configuration
+        - Loads them into a dictionary and returns this dictionary
+
+        Args:
+            pytorch_format: boolean flag to indicate the format of the state dictionary to return.
+
+        Returns:
+            a dictionary with loaded states
+            structure of the returned dictionary:
+            - when pytorch_format is false
+                {
+                    "model":
+                    {
+                        "fp32":
+                        {
+                            model_weight_name:  model_weight_tensor
+                        }
+                    },
+                    "optimizer":
+                    {
+                        model_weight_name:
+                        {
+                            "Moment_1": moment1_tensor,
+                            "Moment_2": moment2_tensor,
+                            "Update_Count": update_tensor # if optimizer is adam, absent otherwise
+                        },
+                        "shared_optimizer_state": # if optimizer is shared, absent otherwise.
+                        {
+                            "step": step_tensor # int array of size 1
+                        }
+                    }
+                    "trainer_options":
+                    {
+                        "mixed_precision": mixed_precision_enabled_bool,
+                        "zero_stage": zero_enabled_stage,
+                        "world_rank": world_rank_int,
+                        "world_size": world_size_int,
+                        "optimizer_name": training_optimizer_name
+                    },
+                    "partition_info":
+                    {
+                        model_weight_name:
+                        {
+                            "original_dim": original_dimension_array
+                        }
+                    }
+                }
+            - when pytorch_format is true
+                {
+                    model_weight_name: model_weight_tensor
+                }
+
+        """
+        if not self._training_session:
+            warnings.warn("ONNX Runtime training session is not initialized yet. " \
+                          "Please run train_step or eval_step at least once before calling ORTTrainer.state_dict().", \
+                          UserWarning)
+            return self._load_state_dict.args[0] if self._load_state_dict else {}
+
+        states = {}
+
+        # load training session model states into the states dictionary
+        self._extract_model_states(states)
+        if pytorch_format:
+            # if pytorch_format is true, return a flat dictionary with only model states
+            # which is compatible with a PyTorch model
+            return states['model']['fp32']
```

Comment:
We don't need to print a warning if mixed precision is enabled because fp32 will be in sync, right?

Reply:
Yes, fp32 and fp16 should always be in sync. So, if the trainer is a mixed precision trainer, the returned fp32 model states should exactly match the fp16 ones (with some tolerable difference). I don't think a warning is necessary here.